### PR TITLE
[WIP] Web push notifications

### PR DESCRIPTION
**Pull request by marcushellberg**
Here is a proof-of-concept implementation of a web push notification service for Vaadin. 

The implementation uses https://github.com/web-push-libs/webpush-java for sending push messages, but if we want to avoid the dependency, we could implement the functionality directly ourselves. See https://golb.hplar.ch/2019/08/webpush-java.html for an example. 

Here is an example app using the API to handle subscriptions and send messages. https://github.com/marcushellberg/vaadin-web-push/blob/using-flow-service/src/main/java/org/vaadin/demo/push/WebPushDemo.java

### CODE REVIEW

**Comment by pleku** [comment]
This would be an interesting addition to V15 / 14.X.

But since the API is not supported on iOS (Safari), I'm not sure if we can include the API directly into `flow-server` as the feature would not work for all end users. Another option would be to ship it as part of a separate artifact (making it an add-on), that you can include when you need the feature.

**Comment by marcushellberg** [comment]
I think it would be good as a separate dependency. Those that don't need the functionality could opt out of it easily. 

We could look into also adding support for Safari, although that will require some extra effort https://developer.apple.com/library/archive/documentation/NetworkingInternet/Conceptual/NotificationProgrammingGuideForWebsites/PushNotifications/PushNotifications.html

**Comment by pleku** [comment]
This PR is getting its one year anniversary 🎂 next Monday so we should get a cake for it ...

If wework on this - it would be a totally _new feature_ - so I want to hear from @mstahv. While it is not a huge amount of work, I think this still requires some documentation and DX testing so I would not try to work on this as _maintenance_. That said, it might be difficult to push this to the roadmap for rest of the year but I would at least give this like two days first and see how far we get with that effort.

Couple of open questions are:
- The changes are done on top of the "old" style bootstrapping and would need to be adopted for the changes done for v15+. The current changes could be a candidate for v14.
- We would need to consider if it is OK to have this just for the server driven views or should we also need to have this for the client driven views, and if so, does it change something. (I'll open a chat about this)

And again, for the server side views I would want to release this as an opt-in feature because there are no signs that Safari would support this.

File: flow-server/src/main/java/com/vaadin/flow/notification/PushNotificationService.java [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+import java.security.Security;
+import java.util.concurrent.CompletableFuture;
+
+public class PushNotificationService {
+
+  private PushService pushService;
+
+  public PushNotificationService(String vapidPubKey, String vapidPrivateKey, String subject) {
+    Security.addProvider(new BouncyCastleProvider());
+    UI.getCurrent().getPage().addJavaScript("frontend://push-notification-helper.js");
```
Comment (pleku) [inline flow-server/src/main/java/com/vaadin/flow/notification/PushNotificationService.java:23]: BTW this helper JS file is missing

**Comment by pleku** [comment]
>We would need to consider if it is OK to have this just for the server driven views or should we also need to have this for the client driven views, and if so, does it change something. (I'll open a chat about this)

After having a discussion on this, it is more clear to me that these changes should also be adopted so that the push notifications can be utilized regardless of whether there is a server driven (Java) or client driven (TS) view. Thus this feature should be _done right_ which means more work but would make it work consistently and for wider use cases. So instead of adopting this contribution, it should be roadmap item that we'll look at it when it becomes a priority.

Until that happens, I think the changes here can be made into an add-on for vaadin.com/directory - I don't see anything that is specifically needed from the framework. The parts in `PwaRegistry` and `BootstrapHandler` should go to a `IndexHtmlRequestListener` (v15+) or `BootstrapListener` that is provided in the add-on and which then the add-on users would register.

Since we don't want to keep PRs open when they are not getting in, I'm closing this one.

**Pull request closed without merging**